beanie/odm/fields.py

```diff
@@ -159,7 +158,16 @@ def __get_pydantic_core_schema__(
         ) -> CoreSchema:  # type: ignore
             return core_schema.json_or_python_schema(
                 python_schema=plain_validator(cls.validate),
-                json_schema=str_schema(),
+                json_schema=plain_validator(
+                    cls.validate,
+                    metadata={
+                        "pydantic_js_input_core_schema": core_schema.str_schema(
+                            pattern="^[0-9a-f]{24}$",
+                            min_length=24,
+                            max_length=24,
+                        )
+                    },
+                ),
```

Comment:
Thank you for your contribution!
Could you please add a test for the generated JSON Schema as well? 
Unfortunately, it is causing issues with FastAPI OpenAPI documentation generation:
```
Cannot generate a JsonSchema for core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})
```

<details>
  <summary>Stack trace</summary>
  
  ```
  .venv/lib/python3.12/site-packages/fastapi/applications.py:981: in openapi
    self.openapi_schema = get_openapi(
.venv/lib/python3.12/site-packages/fastapi/openapi/utils.py:493: in get_openapi
    field_mapping, definitions = get_definitions(
.venv/lib/python3.12/site-packages/fastapi/_compat.py:231: in get_definitions
    field_mapping, definitions = schema_generator.generate_definitions(
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:361: in generate_definitions
    self.generate_inner(schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:540: in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
.venv/lib/python3.12/site-packages/pydantic/main.py:728: in __get_pydantic_json_schema__
    return handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1489: in model_schema
    json_schema = self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1061: in function_before_schema
    return self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1603: in model_fields_schema
    json_schema = self._named_required_fields_schema(named_required_fields)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1393: in _named_required_fields_schema
    field_json_schema = self.generate_inner(field).copy()
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1461: in model_field_schema
    return self.generate_inner(schema['schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:540: in new_handler_func
    json_schema = js_modify_function(schema_or_field, current_handler)
.venv/lib/python3.12/site-packages/beanie/odm/fields.py:178: in __get_pydantic_json_schema__
    json_schema = handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1343: in json_or_python_schema
    return self.generate_inner(schema['json_schema'])
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:565: in generate_inner
    json_schema = current_handler(schema)
.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py:38: in __call__
    return self.handler(core_schema)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:493: in handler_func
    json_schema = generate_for_schema_type(schema_or_field)
.venv/lib/python3.12/site-packages/pydantic/json_schema.py:1088: in function_plain_schema
    return self.handle_invalid_for_json_schema(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pydantic.json_schema.GenerateJsonSchema object at 0x13cbeea20>
schema = {'function': {'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>, 't...e_schema': {'max_length': 24, 'min_length': 24, 'pattern': '^[0-9a-f]{24}$', 'type': 'str'}}, 'type': 'function-plain'}
error_info = "core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})"

    def handle_invalid_for_json_schema(self, schema: CoreSchemaOrField, error_info: str) -> JsonSchemaValue:
>       raise PydanticInvalidForJsonSchema(f'Cannot generate a JsonSchema for {error_info}')
E       pydantic.errors.PydanticInvalidForJsonSchema: Cannot generate a JsonSchema for core_schema.PlainValidatorFunctionSchema ({'type': 'with-info', 'function': <bound method PydanticObjectId.validate of <class 'beanie.odm.fields.PydanticObjectId'>>})
E       
E       For further information visit https://errors.pydantic.dev/2.10/u/invalid-for-json-schema

.venv/lib/python3.12/site-packages/pydantic/json_schema.py:2279: PydanticInvalidForJsonSchema
  ```
  
</details>

Reply:
Hi @ADR-007, this is a closed pull request.
This issue is being addressed in #1080.

Reply:
Thank you!